### PR TITLE
245 improvements in nn module

### LINKED ISSUE #245: Improvements in NN module
Improve NN training pipeline robustness and configurability. 

- `Pipeline` (pyLOM/NN/pipeline.py): Add `evaluate(...)` to run evaluator suites on train/valid/test with optional scaling and verbosity.
- `ClusteredPipeline` (pyLOM/NN/pipeline.py): Allow supplying precomputed model outputs to skip retraining; make classifier/regressor evaluation optional with verbosity control.
- `MLP` (pyLOM/NN/architectures/mlp.py): Support custom losses via `BaseLossFunction`/`TorchLossAdapter`; call dataset train/eval hooks; tougher Optuna search (power-of-two/log suggestions); clearer parameter counting; optional verbose loading.
- `BinaryClassifier` (pyLOM/NN/architectures/binary_classifier.py): Record XGBoost leaf-count as parameter-like metric; allow verbose/silent load.
- `OptunaOptimizer` (pyLOM/NN/optimizer.py): Enable external storage/study reuse; save best parameters with nested keys; standardized optimization reporting.
- `Dataset` (pyLOM/NN/utils.py): Use a factory helper when splitting/filtering so subclass behavior is preserved and easily inheritable.

### PR DESCRIPTION
**Pull request by fermingut**
Closes #245

### CODE REVIEW

**Comment by ArnauMiro** [comment]
@fermingut  since we're into improving is there a chance to move the `Dataset` and `Scalers` into separate files inside `pyLOM/NN/`? One could be `pyLOM/NN/Dataset.py` and the other `pyLOM/NN/Scalers.py`.

**Comment by fermingut** [comment]
@ArnauMiro That sounds good to me. I'll move the `Dataset` and `Scalers` into separate files inside `pyLOM/NN/` now.

File: pyLOM/NN/architectures/mlp.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import os
+import os, math
```
Comment (ArnauMiro) [inline pyLOM/NN/architectures/mlp.py]: why math and not use `np.log2`?

**Review by ArnauMiro — CHANGES_REQUESTED** [review]
Please run the testsuite manually and make sure everything works. Post the output of the testsuite for the NN module, solve the comment and we can merge.

File: pyLOM/NN/architectures/mlp.py (hunk 1 above)
Comment (fermingut) [reply pyLOM/NN/architectures/mlp.py]: Got it, I have updated it

Comment (ArnauMiro) [reply pyLOM/NN/architectures/mlp.py]: @fermingut did you push? I don't see any update

**Comment by fermingut** [comment]
I have run the testsuite using the file "test_suite.json" and everything seems to be ok. This is the output:

Python version 3.10.16 | packaged by conda-forge | (main, Apr  8 2025, 20:53:32) [GCC 13.3.0] 
SINNER=f.gutierrez
DATE=jue 19 mar 2026 13:21:00 CET
HOSTNAME=n008
HOST ARCHITECTURE=x86_64
COMMIT=commit a33ba46a9d88e72084950781216e3a3de8bddbdd
Running <rm -rf /tmp/TESTSUITE/>

                _ ___                /^^\ /^\  /^^\_
    _          _@)@) \            ,,/ '` ~ `'~~ ', `\.
  _/o\_ _ _ _/~`.`...'~\        ./~~..,'`','',.,' '  ~:
 / `,'.~,~.~  .   , . , ~|,   ,/ .,' , ,. .. ,,.   `,  ~\_
( ' _' _ '_` _  '  .    , `\_/ .' ..' '  `  `   `..  `,   \_
 ~V~ V~ V~ V~ ~\ `   ' .  '    , ' .,.,''`.,.''`.,.``. ',   \_
  _/\ /\ /\ /\_/, . ' ,   `_/~\_ .' .,. ,, , _/~\_ `. `. '.,  \_
 < ~ ~ '~`'~'`, .,  .   `_: ::: \_ '      `_/ ::: \_ `.,' . ',  \_
  \ ' `_  '`_    _    ',/ _::_::_ \ _    _/ _::_::_ \   `.,'.,`., \-,-,-,_,_,
   `'~~ `'~~ `'~~ `'~~  \(_)(_)(_)/  `~~' \(_)(_)(_)/ ~'`\_.._,._,'_;_;_;_;_;


Remember: The dragon almost never lies 
Running test AE Cylinder 2D...
**Running <mpirun -np 1  python runs/tsuite_AE2D.py "DATA/CYLINDER.h5" "['VELOX']" "/tmp/TESTSUITE/AE_Cylinder_2D" "{}" | tee /tmp/TESTSUITE/AE_Cylinder_2D.out>
0 Warning! Import - NVTX not present!
  Checking files summary=</tmp/TESTSUITE/AE_Cylinder_2D.log> reference=<references/AE_Cylinder_2D.ref>
  OK
Running test AE Cylinder 2D Fully...
**Running <mpirun -np 1  python runs/tsuite_AE2D_fully.py "DATA/CYLINDER.h5" "['VELOX']" "/tmp/TESTSUITE/AE_Cylinder_2D_fully" "{}" | tee /tmp/TESTSUITE/AE_Cylinder_2D_fully.out>
0 Warning! Import - NVTX not present!
  Checking files summary=</tmp/TESTSUITE/AE_Cylinder_2D_fully.log> reference=<references/AE_Cylinder_2D_fully.ref>
  OK
Running test VAE Cylinder 2D...
**Running <mpirun -np 1  python runs/tsuite_VAE2D.py "DATA/CYLINDER.h5" "['VELOX']" "/tmp/TESTSUITE/VAE_Cylinder_2D" "{}" | tee /tmp/TESTSUITE/VAE_Cylinder_2D.out>
0 Warning! Import - NVTX not present!
  Checking files summary=</tmp/TESTSUITE/VAE_Cylinder_2D.log> reference=<references/VAE_Cylinder_2D.ref>
  OK
Running test VAE Cylinder 2D Fully...
**Running <mpirun -np 1  python runs/tsuite_VAE2D_fully.py "DATA/CYLINDER.h5" "['VELOX']" "/tmp/TESTSUITE/VAE_Cylinder_2D_fully" "{}" | tee /tmp/TESTSUITE/VAE_Cylinder_2D_fully.out>
0 Warning! Import - NVTX not present!
  Checking files summary=</tmp/TESTSUITE/VAE_Cylinder_2D_fully.log> reference=<references/VAE_Cylinder_2D_fully.ref>
  OK
Running test VAE Cylinder 2D fine-tuning...
**Running <mpirun -np 1  python runs/tsuite_VAE2D_finetune.py "DATA/CYLINDER.h5" "['VELOX']" "/tmp/TESTSUITE/VAE_Cylinder_2D_finetune" "{'epochs_ft': 50}" | tee /tmp/TESTSUITE/VAE_Cylinder_2D_finetune.out>
0 Warning! Import - NVTX not present!
  Checking files summary=</tmp/TESTSUITE/VAE_Cylinder_2D_finetune.log> reference=<references/VAE_Cylinder_2D_finetune.ref>
  OK
Skipping test VAE Tensor280 3D...
Running test VAE Channel 3D...
**Running <mpirun -np 1  python runs/tsuite_VAE3D.py "DATA/CHANNEL.h5" "['VELOX']" "/tmp/TESTSUITE/VAE_Channel_3D" "{'nx': 64, 'ny': 64, 'nz': 64}" | tee /tmp/TESTSUITE/VAE_Channel_3D.out>
0 Warning! Import - NVTX not present!
  Checking files summary=</tmp/TESTSUITE/VAE_Channel_3D.log> reference=<references/VAE_Channel_3D.ref>
  OK
Running test MLP airfoil data...
**Running <mpirun -np 1  python runs/tsuite_MLP.py "DATA/AIRFOIL.h5" "['cp']" "/tmp/TESTSUITE/MLP_airfoil" "{}" | tee /tmp/TESTSUITE/MLP_airfoil.out>
0 Warning! Import - NVTX not present!
0 Warning! The model has been loaded with weights_only set to False. According with torch documentation, this is not recommended if you do not trust the source of your saved model, as it could lead to arbitrary code execution.
  Checking files summary=</tmp/TESTSUITE/MLP_airfoil.log> reference=<references/MLP_airfoil.ref>
  OK
Running test KAN airfoil data...
**Running <mpirun -np 1  python runs/tsuite_KAN.py "DATA/AIRFOIL.h5" "['cp']" "/tmp/TESTSUITE/KAN_airfoil" "{}" | tee /tmp/TESTSUITE/KAN_airfoil.out>
0 Warning! Import - NVTX not present!
0 Warning! The model has been loaded with weights_only set to False. According with torch documentation, this is not recommended if you do not trust the source of your saved model, as it could lead to arbitrary code execution.
  Checking files summary=</tmp/TESTSUITE/KAN_airfoil.log> reference=<references/KAN_airfoil.ref>
  OK
Skipping test PINN Burgers equations...

                ^    ^
               / \  //\ 
    |\___/|      /   \//  .\ 
    /O  O  \__  /    //  | \ \ 
    /     /  \/_/    //   |  \  \ 
    @___@'    \/_   //    |   \   \ 
    |       \/_ //     |    \    \ 
    |        \///      |     \     \ 
    _|_ /   )  //       |      \     _\ 
    '/,_ _ _/  ( ; -.    |    _ _\.-~        .-~~~^-.
    ,-{        _      `-.|.-~-.           .~         `.
    '/\      /                 ~-. _ .-~      .-~^-.  \ 
     `.   {            }                   /      \  \ 
    .----~-.\        \-'                 .~         \  `. \^-.
    ///.----..>    c   \             _ -~             `.  ^-`   ^-_
    ///-._ _ _ _ _ _ _}^ - - - - ~                     ~--,   .-~
            Some of the tests are OK...                   /.-'

test <VAE_Tensor280_3D>: VAE Tensor280 3D SKIPPED!
test <PINN_burgers>: PINN Burgers equations SKIPPED!

File: pyLOM/NN/architectures/mlp.py (hunk 1 above)
Comment (fermingut) [reply pyLOM/NN/architectures/mlp.py]: @ArnauMiro I was running the testsuite sorry, i have pushed it now

**Review by ArnauMiro — APPROVED** [review]
